### PR TITLE
lxd: Replace gorilla/mux with stdlib http.ServeMux

### LINKED ISSUE #16587: Replace `gorilla/mux` by `http.ServeMux`
### Please confirm

- [x] I have searched existing issues to check if an issue already exists for my feature request.

### Is your feature request related to a problem? Please describe.

The stdlib `http.ServeMux` has everything needed to replace `gorilla/mux` which would remove a dependency on a third party library.

### Describe the solution you'd like

One less 3rd party dependency.

### Describe alternatives you've considered

_No response_

### Additional context

https://github.com/canonical/lxd/pull/16367 should deal with the `lxd-agent` part after cherry-picking some commits from Incus.

Incus folks are trying to do the same for the whole project but it's still in draft: https://github.com/lxc/incus/pull/2451

### PR DESCRIPTION
**Pull request by larrasket**
Go 1.22+ added path variables and wildcard support to http.ServeMux,
   making gorilla/mux redundant. This removes the dependency.

One complication: ServeMux panics when `/images/aliases/{name...}` and `/images/{fingerprint}/export` are both registered, since both match `/images/aliases/export`. Image aliases can contain slashes, so the wildcard is necessary. `lxdMux` (a thin wrapper around `ServeMux`) handles `{var...}` patterns via prefix matching before delegating to `ServeMux`.  

Behavioral notes:
- `ServeMux` cleans paths (`//` → `/`) unlike gorilla's `SkipClean(true)`
- `r.PathValue()` returns decoded values; existing `url.PathUnescape()` calls become no-ops but are harmless
 
Close #16587 

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/canonical/lxd/blob/main/CONTRIBUTING.md) and attest that all commits in this PR are [signed off](https://github.com/canonical/lxd/blob/main/CONTRIBUTING.md#including-a-signed-off-by-line-in-your-commits), [cryptographically signed](https://github.com/canonical/lxd/blob/main/CONTRIBUTING.md#commit-signature-verification), and follow this project's [commit structure](https://github.com/canonical/lxd/blob/main/CONTRIBUTING.md#commit-structure).
- [x] I have checked and added or updated relevant documentation.

### CODE REVIEW

**Review by tomponline — CHANGES_REQUESTED** [review]
Please can this be split into logical commits rather than 1 massive one. Thanks

I suggest at least one commit per package.

File: shared/context.go [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package shared
```
Comment (tomponline) [inline shared/context.go]: If these variables are not needed by the client then they should not be in `shared` package.

**Comment by simondeziel** [comment]
> Image aliases can contain slashes

Hmm, good catch! That's surprising and indeed it works. I don't like it :/

**Comment by larrasket** [comment]
Yeah I was surprised too when I saw the test. I’m unsure whether changing it will be a breaking change.

**Comment by simondeziel** [comment]
> Yeah I was surprised too when I saw the test. I’m unsure whether changing it will be a breaking change.

It'd indeed is be a breaking change but the affected population is likely small. Let's hear what @tomponline has to say about it.

**Comment by tomponline** [comment]
> Image aliases can contain slashes, so the wildcard is necessary.

But they should be path encoded right?

**Comment by tomponline** [comment]
If the path encoding/decoding is handled correctly (and it needs to be that way), it shouldnt matter if image aliases can contain slashes. If it is causing problems thats a sign the path encoding/decoding isn't correct I suspect.

File: shared/context.go (hunk 1 above)
Comment (larrasket) [reply shared/context.go]: I've moved it to lxd/request/const.go alongside the other CtxKey constants

File: client/request.go [hunk 2]
```diff
@@ -56,5 +58,14 @@ func NewRequestWithContext(ctx context.Context, method string, url string, data
 		req.Header.Set("If-Match", ETag)
 	}
 
+	// Copy path values stored in the context by lxdMux so they survive
+	// internal request forwarding (e.g., devLXD → main API handlers).
```
Comment (simondeziel) [inline client/request.go]: Coding agents love to use `→` it seems but I find them annoying to work with. Would you mind using `->` or similar arrow-emulation in ASCII ;)

File: lxd/images.go [hunk 3]
```diff
@@ -111,7 +110,7 @@ var imageAliasesCmd = APIEndpoint{
 }
 
 var imageAliasCmd = APIEndpoint{
-	Path:        "images/aliases/{name:.*}",
+	Path:        "images/aliases/{name}",
 	MetricsType: entity.TypeImage,
```
[the remark this answers is not in the document]
Comment (larrasket) [reply lxd/images.go:114]: resolving this as the behavioral change was intended.  clients should URL-encode slashes as %2F, which the client library is already doing.

**Comment by tomponline** [comment]
Please can you rebase

File: shared/context.go (hunk 1 above)
Comment (tomponline) [reply shared/context.go]: It still seems to be in shared?

Comment (larrasket) [reply shared/context.go]: the client’s NewRequestWithContext now copies mux path vars from the context back onto the request

File: lxd/storage_volumes.go [hunk 4]
```diff
@@ -1049,9 +1048,8 @@ func storagePoolVolumesPost(d *Daemon, r *http.Request) response.Response {
 	}
 
 	// Handle being called through the typed URL.
-	_, ok := mux.Vars(r)["type"]
-	if ok {
-		req.Type, err = url.PathUnescape(mux.Vars(r)["type"])
+	if r.PathValue("type") != "" {
+		req.Type, err = url.PathUnescape(r.PathValue("type"))
```
Comment (simondeziel) [inline lxd/storage_volumes.go]: Please cache `r.PathValue("type")` in a local variable instead of repeating the call.

File: lxd/api.go [hunk 5]
```diff
@@ -281,8 +277,153 @@ func metricsServer(d *Daemon) *http.Server {
...
+
+		r.SetPathValue(ph.varName, decoded)
+		ph.handler(w, r)
+
+		return
+	}
+
+	// a legit 404.
+	m.ServeMux.ServeHTTP(w, r)
+}
```
Comment (simondeziel) [inline lxd/api.go:241]: This should reduce the amount of allocations:
```suggestion
pathLen := len(escapedPath)

for _, ph := range m.prefixHandlers {
    preLen := len(ph.prefix)
    sufLen := len(ph.suffix)

    if pathLen < preLen+sufLen || !strings.HasPrefix(escapedPath, ph.prefix) {
        continue
    }

    remainder := escapedPath[preLen:]
    if !strings.HasSuffix(remainder, ph.suffix) {
        continue
    }

    val := remainder[:len(remainder)-sufLen]

    if val == "" || strings.Contains(val, "/") {
        continue
    }
    // ... rest of logic
}
```

**Comment by simondeziel** [comment]
@larrasket do you still plan to move this forward... that'd be a very nice contribution to land, seriously ;)

**Comment by larrasket** [comment]
Hi @simondeziel, yes I do. Just been busy with Eid, will update this soon.

—
Regards
Salih

File: lxd/daemon.go
Comment (simondeziel) [inline lxd/daemon.go]: I think this became dead, isn't it?

Comment (simondeziel) [inline lxd/daemon.go]: Same as `Name`, I think this is dead now.

Comment (simondeziel) [inline lxd/daemon.go]: Probably dead code too.

File: lxd/api.go [hunk 6]
```diff
@@ -291,8 +287,162 @@ func metricsServer(d *Daemon) *http.Server {
...
+		// Parse the pattern to extract prefix, variable name, and suffix.
+		// For example, "images/{fingerprint}/export" becomes:
+		// prefix="/1.0/images/", varName="fingerprint", suffix="/export".
+		start := strings.Index(pattern, "{")
+		end := strings.Index(pattern, "}")
+		if start == -1 || end == -1 || end <= start {
+			panic(fmt.Sprintf("Conflicting route pattern without parsable path variable: %q", pattern))
+		}
+
+		m.prefixHandlers = append(m.prefixHandlers, lxdPrefixHandler{
```
Comment (simondeziel) [inline lxd/api.go:143]: Would it make sense to defend against duplicated `HandleFunc` registration of the same path but with different handlers?

Something like this?
```
        prefix := pattern[:start]
        suffix := pattern[end+1:]

        // Guard against duplicate registrations: a second call with the same
        // conflicting pattern would append a second unreachable entry and the
        // first registered handler would always win in ServeHTTP.
        for _, existing := range m.prefixHandlers {
            if existing.prefix == prefix && existing.suffix == suffix {
                panic(fmt.Sprintf("Duplicate conflicting route registration for pattern %q", pattern))
            }
        }
```

File: lxd/api.go [hunk 7]
```diff
@@ -291,8 +287,162 @@ func metricsServer(d *Daemon) *http.Server {
...
+		// ServeMux panics with a message containing "conflicts with pattern"
+		// when two patterns overlap. I tested this with Go 1.22-1.26.  If Go
+		// changes the wording, tryRegister will re-panic (safe failure) instead
+		// of silently dropping routes.
+		if strings.Contains(msg, "conflicts with pattern") {
+			registered = false
+			return
+		}
+
+		panic(r)
```
Comment (simondeziel) [inline lxd/api.go]: Would it be possible to actually panic on duplicated registrations? Go panic message for `ServeMux` pattern conflicts should includes both conflicting patterns in quotes:

`pattern "/1.0/images/{fingerprint}" (registered at ...) conflicts with pattern "/1.0/images/{fingerprint}" (registered at ...)`

Something along those lines, maybe?


```suggestion
    // Panic on unrelated panic or changed wording.
    if !strings.Contains(msg, "conflicts with pattern") {
        panic(r)
    }

    // Distinguish a legitimate overlap (two different patterns that
    // conflict) from an accidental exact duplicate. Go's panic message
    // for ServeMux conflicts is of the form:
    //  `pattern "A" ... conflicts with pattern "B" ...`
    // If the quoted pattern appears twice in the message, A == B:
    // the caller registered the exact same pattern twice, re-panic
    // rather than silently dropping the duplicate.
    quoted := fmt.Sprintf("%q", pattern)
    if strings.Count(msg, quoted) >= 2 {
        // Both conflicting patterns are identical to ours: exact duplicate.
        panic(r)
    }

    registered = false
```

File: lxd/daemon.go
Comment (larrasket) [reply lxd/daemon.go]: You mean name? No, it's used in `instances.go` and `instance_logs.go`.

Comment (larrasket) [reply lxd/daemon.go]: Used in api.go:218.

Comment (larrasket) [reply lxd/daemon.go]: Same

Comment (simondeziel) [reply lxd/daemon.go]: True but I also think this `api.go:218` is dead code too.

With the removal of `/1.0/containers` and `/1.0/virtual-machines`, I think we lost the last API endpoint aliases.

Comment (simondeziel) [reply lxd/daemon.go]: True, those files populate the field, but the value only flows into `route.Name(c.Name)` in `gorilla/mux` (`daemon.go:1021`), which registers a named route for URL reversal. Since `router.Get(name)` and `mux.CurrentRoute(r).GetName()` are never called anywhere in the codebase, that's a no-op even today. The field is already dead code, independent of the `http.ServeMux` migration.

File: test/lint/bin-size.sh [hunk 8]
```diff
@@ -12,8 +12,8 @@ echo "Check binaries size limits"
 
 # bin/max (sizes are in MiB)
 declare -rA sizes=(
-    ["lxc"]=17
-    ["lxd-agent"]=14
+    ["lxc"]=15
+    ["lxd-agent"]=13
```
Comment (simondeziel) [inline test/lint/bin-size.sh]: This is the biggest jump I've seen ;)
Well done!

Comment (larrasket) [reply test/lint/bin-size.sh]: Thanks. There's a catch though, building it locally results on ~14.8MiB binary, however, the CI builds are exceeding 16 still. My guess is that the effect on amd64 is smaller than my arm64 (I confirmed by a cross build)

Comment (simondeziel) [reply test/lint/bin-size.sh]: Ah, interesting data point. Still going in the right direction :)

**Review by simondeziel — DISMISSED** [review]
LGTM, thanks!

File: lxd/api.go [hunk 9]
```diff
@@ -116,8 +112,153 @@ func metricsServer(d *Daemon) *http.Server {
...
+	// Collect path variable names from this pattern.
+	var varNames []string
+	for _, seg := range strings.Split(pattern, "/") {
+		if len(seg) > 2 && seg[0] == '{' && seg[len(seg)-1] == '}' && seg != "{$}" {
+			name := strings.TrimSuffix(seg[1:len(seg)-1], "...")
+			varNames = append(varNames, name)
+		}
+	}
+
+	// Wrap handler to re-encode path values and store them in context.
```
Comment (tomponline) [inline lxd/api.go]: This seems to be some extra complexity due to:

"handlers expect encoded values (gorilla/mux UseEncodedPath behavior) and call url.PathUnescape"

But this isn't required, ultimately (I understand anyway) LXD handlers just need to ensure that they get the relevant path segment made available to them unescaped, and we need to ensure that the mux doesn't treat an escaped `/` in a path escaped segment as part of the path.

E.g.

`/1.0/something/foo%2Fsnap0`

shouldn’t be treated as

`/1.0/something/foo/snap0`

But rather make the last segment available to the handler as:

`foo/snap0`

File: lxd/api.go [hunk 10]
```diff
@@ -116,8 +112,153 @@ func metricsServer(d *Daemon) *http.Server {
...
+				v := r.PathValue(name)
+				if v != "" {
+					encoded := url.PathEscape(v)
+					r.SetPathValue(name, encoded)
+					vars[name] = encoded
+				}
+			}
+
+			if len(vars) > 0 {
+				r = r.WithContext(context.WithValue(r.Context(), shared.CtxMuxPathVars, vars))
```
Comment (tomponline) [inline lxd/api.go]: I'd like us to avoid having to put more info about the request in the context, this shouldn't be needed if the mux is providing us what we need, and if its not then maybe we shouldn't be switching to this package.

**Review by tomponline — CHANGES_REQUESTED** [review]
As aliases are no longer used, I think we can start with a standalone commit to remove support for them.

**Comment by tomponline** [comment]
Thanks @larrasket 

I'm a bit concerned that this PR seems to include some extra complexity that is unnecessary.

I don't understand why we need to include path escaped segment values in the request context and them restore them later, only to be path unescaped inside the handler.

Can we avoid this and use the mux as it is intended?

**Comment by larrasket** [comment]
I agree. When I was migrating I noticed that the stdlib's `r.PathValue()` returns decoded values. tests were not working because handlers called `PathUnescape` on already-decoded values, which was a no-op for most strings but would break edge cases as mentioned before, the fix was to er-encode. I'm working on removing this hack already.

File: lxd/api.go [hunk 11]
```diff
@@ -116,8 +101,147 @@ func metricsServer(d *Daemon) *http.Server {
...
+	varName string
+	suffix  string
+	handler http.HandlerFunc
+}
+
+// HandleFunc registers a handler for the given pattern. If the pattern conflicts
+// with an already-registered ServeMux pattern, it falls back to prefix-based matching.
+func (m *lxdMux) HandleFunc(pattern string, handler func(http.ResponseWriter, *http.Request)) {
+	if !m.tryRegister(pattern, handler) {
+		// Parse the pattern to extract prefix, variable name, and suffix.
```
Comment (tomponline) [inline lxd/api.go:123]: I wonder if we should avoid this complexity and instead use the stdlib mux as it was intended, and handle the ambiguous endpoints inside a single handler function.

I.e have `images/{fingerprint}/export` handle the case where fingerprint is "alias" and then call the `images/aliases/{name}` hander?

Comment (larrasket) [reply lxd/api.go:123]: I don't think this will work. `imageAliasCmd` already handles 5 methods on `images/aliases/{name}`, and we have `images/{fingerprint}/[export-GET][refresh-[POST][secret-GET]` conflicting. Even if I move the dispatch logic to the three handlers I'd only over GET and POST on aliases named "export", "refresh", "secret" and other methods (DELETE, PATCH, PUT) on an alias named "export" will have no handler and they'll be failing to 404. 
for example imagine where DELETE /1.0/images/aliases/export would go (assuming we have an alias named export)

Comment (larrasket) [reply lxd/api.go:123]: @tomponline friendly ping

Comment (MusicDin) [reply lxd/api.go:123]: Huh, yeah the paths indeed conflict, and stdlib mux would panic in such case.
To be sincere, I would rather see those routes fixed in the future as they should not conflict in the first place instead of looking for workarounds.

Comment (tomponline) [reply lxd/api.go:123]: I'm thinking we stick with gorilla mux for now and then when we can make breaking API changes we can switch to go's mux then.

Comment (tomponline) [reply lxd/api.go:123]: @larrasket 

> I don't think this will work. imageAliasCmd already handles 5 methods on images/aliases/{name}, and we have images/{fingerprint}/[export-GET][refresh-[POST][secret-GET] conflicting. Even if I move the dispatch logic to the three handlers I'd only over GET and POST on aliases named "export", "refresh", "secret" and other methods (DELETE, PATCH, PUT) on an alias named "export" will have no handler and they'll be failing to 404.
for example imagine where DELETE /1.0/images/aliases/export would go (assuming we have an alias named export)

I don't really follow your point here.

Why can't we have the `"images/{fingerprint}/export",` handler detect if the `{fingerprint}` is "aliases" and then call the image aliases hander?

And for the unsupported verbs, like DELETE, we can have either handler respond with a 404 as you say?

Comment (larrasket) [reply lxd/api.go:123]: first, stdlib ServeMux panics at registration when two patterns overlap and neither is a subset of the other. `images/{fingerprint}/export` and`images/aliases/{name}` are exactly that case (`images/aliases/export` matches both, neither is more specific). So I can't register both and have one dispatch into the other,I have to drop one.

Now, If I drop `images/aliases/{name}` and keep only `images/{fingerprint}/{export,secret,refresh}`, then `GET /1.0/images/aliases/myalias` (where the alias name isn't one of those three suffixes) has no route at all and 404s. That breaks the alias API in general, not just the DELETE-on-alias-named-export case I mentioned, which is what I was trying to get at before.

The way to use stdlib mux as intended would be to collect the three suffix endpoints into a single`images/{fingerprint}/{action}` pattern and dispatch on action inside one handler.`images/aliases/{name}` is (then) a proper subset of that and stdlib mux resolves it cleanly. Happy to do that as a follow-up if you'd prefer it over lxdPrefixHandler, but it's a similar amount of dispatch logic, just inside the handler instead of inside the mux.

Comment (tomponline) [reply lxd/api.go:123]: Hi @larrasket I was using Claude Sonnet 4.5 to try and solve this issue and it came up with https://github.com/canonical/lxd/pull/18309 (with a fair bit of coaxing) to use the delegated handler approach we discussed.

I still need to go through it in detail, but I like how it added the new `EndpointResolver` concept.

Interested to know your thoughts on it.
